experimental/SetPartitions/src/GenerateCategory.jl

```diff
@@ -0,0 +1,464 @@
+- `tracing`: optinal input: activate tracing and get the output (category, trace)
+- `max_artifical`: optional input: allow partitions to grow greater while construction process
+
+# Returns
+- list of all partitions size n constructed from partitions in p
+
+# Examples
+```julia-repl
+julia> length(construct_category([Partition([1, 2], [2, 1])], 6))
+105
+julia> length(construct_category([Partition([1, 2], [2, 1])], 6, true))
+[<Partition([1, 2], [2, 1])> ∩ P(6), Dict{Partition -> Tuple}]
+```
+"""
+function construct_category(p::Array, n::Int, tracing::Bool = false, max_artificial::Int = 0, spatial_rotation::Function = nothing_function() = nothing)
+    
+    @assert length(methods(spatial_rotation)[1].sig.parameters)-1 == 0 || (p isa Array{SpatialPartition})
+
+    # store all candidates found
+    all_partitions::Set{AbstractPartition} = Set()
+
+    # all candidates stored in dict from size to partition
+    all_partitions_by_size::Dict{Int, Set{AbstractPartition}} = Dict()
+
+    # all candidates stored in tuple with a dict for top and bottom size
+    all_partitions_by_size_top_bottom::Array = [Dict(), Dict()]
+
+    # store partitions already unary
+    already_u::Set{AbstractPartition} = Set()
+
+    # store partitions already tensor product
+    already_t::Set{Array{AbstractPartition}} = Set()
+
+    # store partitions already composition
+    already_c::Set{Array{AbstractPartition}} = Set()
+
+    # end output: All partitions found of size n
+    all_partitions_of_size_n::Array{AbstractPartition} = []
+
+    # all candidates for unary operations
+    to_unary::Set{AbstractPartition} = Set(copy(p))
+
+    # trace for tracing
+    trace::Dict = Dict()
+
+    # compare allowed expansion size with max(n, max_length)
+    max_length::Int = max(n, maximum(size(i) for i in p))
+
+    if max_artificial > 0
+        max_length = max(max_length, max_artificial)
+    end
+
+    # define for all i <= size an empty set in which we fill the corresponding partition of size i (for tensor)
+    for i in 0:max_length
+        all_partitions_by_size[i] = Set()
+    end
+
+    # define for all bottom and top size an empty set in which we fill the corresponding partition
+    all_partitions_by_size_top_bottom[1] = Dict(i => Set() for i in 0:max_length)
+    all_partitions_by_size_top_bottom[2] = Dict(i => Set() for i in 0:max_length)
+
+    # add all partitions in p to all_partitions_by_size and all_partitions_by_size_top_bottom
+    tuple_list_all_partitions::Array = []
+    for i in all_partitions
+        push!(tuple_list_all_partitions, i)
+    end
+    for i in vcat(p, tuple_list_all_partitions)
+        all_partitions_by_size = add_partition_to_dict(all_partitions_by_size, i)
+        all_partitions_by_size_top_bottom = add_partition_to_composition_dict(all_partitions_by_size_top_bottom, i)
+    end
+    
+    # while new were found apply on them unary tensor and composition
+    stop_whole::Bool = false
```

Comment:
This way of writing code with a variable specifier is rather unidiomatic.
```suggestion
    stop_whole = false
```